### PR TITLE
pAI Encryption Keys

### PR DESCRIPTION
**Pull request by Geevies**
* Clicking on a folded up pAI with an encryption key now slots the key into the pAI card, granting it access to those radio channels.
* You can remove pAI encryption keys by using a screwdriver on them.

image
image

### CODE REVIEW

**Comment by Geevies** [comment]
!review

**Review by Chada-1 — APPROVED** [review]

**Review by Arrow768 — DISMISSED** [review]
Needs to conform to the existing mechanics for encryption keys:

The key needs to be installed physically and may not be "copied"

**Comment by Geevies** [comment]
!review

File: code/game/objects/items/devices/paicard.dm [hunk 1]
```diff
@@ -47,7 +48,60 @@
...
+		var/added_channels = FALSE
+		for(var/thing in EK.channels)
+			if(!radio.channels[thing])
+				added_channels = TRUE
+				break
+		if(!added_channels)
+			for(var/thing in EK.additional_channels)
+				if(!radio.channels[thing])
+					added_channels = TRUE
+					break
```
Comment (mikomyazaki) [inline code/game/objects/items/devices/paicard.dm:65]: You can do both of these at once like:
```suggestion
		for(var/thing in EK.channels | EK.additional_channels)
			if(!radio.channels[thing])
				added_channels = TRUE
				break
```

If you want to.

File: code/game/objects/items/devices/paicard.dm [hunk 2]
```diff
@@ -47,7 +48,60 @@
...
+	radio.channels = list()
+
+	for(var/keyslot in installed_encryptionkeys)
+		var/obj/item/device/encryptionkey/EK = keyslot
+
+		for(var/ch_name in EK.channels)
+			radio.channels[ch_name] = radio.FREQ_LISTENING
+
+		for(var/ch_name in EK.additional_channels)
+			radio.channels[ch_name] = radio.FREQ_LISTENING
```
Comment (mikomyazaki) [inline code/game/objects/items/devices/paicard.dm:96]:
```suggestion
		for(var/ch_name in EK.channels | EK.additional_channels) 
			radio.channels[ch_name] = radio.FREQ_LISTENING
```

Same here.

File: code/modules/modular_computers/file_system/programs/pai/radio.dm [hunk 3]
```diff
@@ -37,11 +37,15 @@
...
 	for(var/ch_name in host.radio.channels)
-		var/ch_stat = host.radio.channels[ch_name]
-		VUEUI_SET_CHECK(data["channels"][ch_name], !!(ch_stat & host.radio.FREQ_LISTENING), ., data)
-	
+		var/list/channel_info = list(
+			"name" = ch_name,
+			"listening" = (host.radio.channels[ch_name] & host.radio.FREQ_LISTENING)
+		)
+		pai_channels[++pai_channels.len] = channel_info
+	VUEUI_SET_CHECK(data["channels"], pai_channels, ., data)
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/pai/radio.dm:47]: `VUEUI_SET_CHECK` does not work for lists. You must use `VUEUI_SET_CHECK_LIST` instead.

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/modules/modular_computers/file_system/programs/pai/radio.dm [hunk 4]
```diff
@@ -44,7 +44,7 @@
 			"listening" = (host.radio.channels[ch_name] & host.radio.FREQ_LISTENING)
 		)
 		pai_channels[++pai_channels.len] = channel_info
-	VUEUI_SET_CHECK(data["channels"], pai_channels, ., data)
+	data["channels"] = pai_channels
```
Comment (mikomyazaki) [inline code/modules/modular_computers/file_system/programs/pai/radio.dm:47]: If you're going to do this, you need to always return data. It won't return data currently if *only* pai_channels changes.
Recommend removing the VUEUI_SET_CHECKS and just doing `return data` at the end.

Technically we are already always returning data already due to how header data works (which is bad but we've got it *everywhere* now...)

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

**Review by mikomyazaki — APPROVED** [review]

**Pull request merged**